contracts/src/Tracker.sol

```diff
@@ -93,39 +94,44 @@ contract Tracker is EIP712 {
 
         digestUsed[digest] = true;
         checkIns[challengeId][msg.sender].push(timestamp);
+
         if (checkIns[challengeId][msg.sender].length == challenges[challengeId].minimunCheckIns) {
-            challenges[challengeId].totalStake -= challenges[challengeId].stakePerUser;
-            balances[msg.sender] += challenges[challengeId].stakePerUser;
             succeedUsers[challengeId].push(msg.sender);
+            claimable[challengeId][msg.sender] = true;
         }
         emit CheckIn(msg.sender, challengeId, timestamp);
     }
 
     // settle a challenge after the ending timestamp
     function settle(uint256 challengeId) public {
         //require(block.timestamp > challenges[challengeId].endTimestamp, "Challenge has not ended");
+        require(!challenges[challengeId].settled, "challenge already settled");
         emit Settle(challengeId);
         challenges[challengeId].settled = true;
+        uint256 succeedUserCounts = succeedUsers[challengeId].length;
 
-        if (succeedUsers[challengeId].length == 0 || challenges[challengeId].totalStake == 0) return;
+        uint256 halfFailedUserStake =
+            (challenges[challengeId].totalStake - (succeedUserCounts * challenges[challengeId].stakePerUser)) / 2;
+        challenges[challengeId].totalStake -= halfFailedUserStake;
+        IERC20(underlyingToken).safeTransfer(challenges[challengeId].donateDestination, halfFailedUserStake);
+    }
 
-        uint256 halfStakeBalance = challenges[challengeId].totalStake / 2;
-        uint256 bonus = halfStakeBalance / succeedUsers[challengeId].length;
+    function getClaimableAmount(uint256 challengeId, address user) public view returns (uint256) {
+        uint256 succeedUserCount = succeedUsers[challengeId].length;
 
-        for (uint256 i = 0; i < succeedUsers[challengeId].length; i++) {
-            address user = users[challengeId][i];
-            balances[user] += bonus;
+        if (!challenges[challengeId].settled || !claimable[challengeId][user] || succeedUserCount == 0) {
+            return 0;
         }
 
-        IERC20(underlyingToken).safeTransfer(challenges[challengeId].donateDestination, halfStakeBalance);
+        return challenges[challengeId].totalStake / succeedUserCount;
     }
 
-    //FIXME: This should be withdraw by challenge id
-    function withdraw() public {
-        uint256 balance = balances[msg.sender];
-        require(balance > 0, "Insufficient balance");
-        balances[msg.sender] = 0;
-        IERC20(underlyingToken).safeTransfer(msg.sender, balance);
+    function withdraw(uint256 challengeId) public {
+        require(challenges[challengeId].settled, "challenge not yet settled");
+        require(claimable[challengeId][msg.sender], "user not eligible or already claimed");
```

Comment:
I don't think we have to check here: if claimable is false, `getClaimableAmount` is going to be 0. The check is duplicated, can just throw on amount == 0.